Task/69113 keep openapi explorer up to date

toy opened the pull request:
# Ticket
https://community.openproject.org/wp/69113

# What are you trying to accomplish?
* Update openapi-explorer
* Fix problem with base tag: openapi-explorer uses `pushState` with anchor only, but this doesn't work with base tag, as anchor is added to the base url instead of current url
* Fix double scroll on api/docs page

# What approach did you choose and why?
* Use openapi-explorer versioned through package.json (and separately bundled) instead of old version in vendor dir
* To fix conflict with base tag:
  * At first two very hacky solutions:
    * https://github.com/opf/openproject/pull/21060
    * https://github.com/opf/openproject/pull/21061
  * Instead in openapi-explorer - https://github.com/Authress-Engineering/openapi-explorer/pull/292
* Hacky fix for double scroll by setting openapi explorer parent tag height and width explicitly to prevent the script from forcing 100vw/100vh causing double scrollbars
* Also use `RAILS_RELATIVE_URL_ROOT` environment variable for dev frontend server (`npm run serve`) so it affects not only backend paths, but also frontend ones

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

toy committed:
set serve prifix of angular dev server to RAILS_RELATIVE_URL_ROOT

toy committed:
[#69113] Keep OpenAPI Explorer up to date

https://community.openproject.org/work_packages/69113

toy committed:
test that api docs page keeps the path

toy committed:
update openapi-explorer to 2.4.784 including it through packages

toy committed:
hack adapted from #21061 to enhance history.pushState

openapi-explorer uses pushState with anchor only, but this doesn't work
with base tag, as anchor is added to the base url instead of current
url. So override the replaceState function to construct the full url
based on current location

Probably not needed very soon, as
Authress-Engineering/openapi-explorer#292 was quickly merged

toy committed:
prevent openapi-explorer parent from getting explicit width and height from the script

toy committed:
use openapi-explorer version working with base tag

Authress-Engineering/openapi-explorer#292

cbliard reviewed (approved):
Nice. It's much better than the 2 previous hacks. Thanks for pushing this forward relentlessly.

One thing that seems to not work well is the back button after clicking an OpenProject link. Here is a repro:
- open `/api/docs`
- click OpenProject logo in top bar
- click back
  - expected: go back to api docs page
  - actual: the url is updated, but the page content is not changed.
  - fun thing: if reloading the page *before* using the back button, it works as expected (so I think it's related to turbo).

But that should not prevent from merging.

There is also an issue with back button inside the api docs page: it seems to store the same url twice. Here is a repro:
- open `/api/docs`
- click some links from the left menu
- click back multiple times
  - expected: go back each time back button is pressed
  - actual: have to click twice to get the url updated
  - it was already doing it before, where the url would be `/#?route=overview--introduction` on first click, then `/api/docs#?route=overview--introduction` on second click, and so on.

Same, still good to merge IMHO because the behavior is so much better than before.

toy committed:
disable turbo for links from api docs page

Otherwise navigating back to docs page fails

toy committed:
move check for api doc enabledness to before action

Otherwise there is double render happending and it is cleaner to use a
before action than to return when rendirng 404

toy commented:
@cbliard Thank you for review and nice finds!

First problem I resolved by disabling turbo on api docs page, so clicking menu links skips turbo which leads to also skipping turbo when navigating back (05f5b9474f6896c6d28751cccba25ee962e405b6, includes a test)

Second problem is ~~in openapi-explorer => https://github.com/Authress-Engineering/openapi-explorer/pull/293~~ fixed (c05ed7d93facdc12555292ebc4ef49e4eb589c21)

toy committed:
use openapi-explorer version with a fix for history

Fir for history: Authress-Engineering/openapi-explorer#293

Discovered by @cbliard
https://github.com/opf/openproject/pull/21484#pullrequestreview-3597442766

toy committed:
deduplicate test textx in api docs specs

cbliard commented:
Nice one 👏
